subxt/src/backend/utils.rs

```diff
@@ -0,0 +1,291 @@
+                    Err(err) => {
+                        if !err.is_disconnected_will_reconnect() {
+                            NextState::Closed(Some(err))
+                        } else {
+                            NextState::Resubscribe {
+                                pending: (self.resubscribe)(),
+                                err,
+                            }
+                        }
+                    }
+                },
+                PendingOrStream::Closed => NextState::Closed(None),
+            };
+
+            match next_state {
+                NextState::Resubscribe { pending, err } => {
+                    self.state = PendingOrStream::Pending(pending);
+                    return Poll::Ready(Some(Err(err)));
+                }
+                NextState::Stream(stream) => {
+                    self.state = PendingOrStream::Stream(stream);
+                }
+                NextState::Closed(maybe_err) => {
+                    self.state = PendingOrStream::Closed;
+
+                    if let Some(err) = maybe_err {
+                        return Poll::Ready(Some(Err(err)));
+                    } else {
+                        return Poll::Ready(None);
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Retry a future until it doesn't return a disconnected error.
+pub async fn retry<T, F, R>(mut retry_future: F) -> Result<R, Error>
+where
+    F: FnMut() -> T,
+    T: Future<Output = Result<R, Error>>,
+{
+    const REJECTED_MAX_RETRIES: usize = 10;
+    let mut rejected_retries = 0;
+
+    loop {
+        match retry_future().await {
+            Ok(v) => return Ok(v),
+            Err(e) => {
+                if e.is_disconnected_will_reconnect() {
+                    continue;
+                }
+
+                // TODO: https://github.com/paritytech/subxt/issues/1567
+                // This is a hack because if a reconnection occurs
+                // the order of pending calls is not guaranteed.
+                //
+                // Such that it's possible the a pending future completes
+                // before `chainHead_follow` is established with fresh
+                // subscription id.
+                //
+                if e.is_rejected() && rejected_retries < REJECTED_MAX_RETRIES {
+                    rejected_retries += 1;
+                    continue;
+                }
+
+                return Err(e);
+            }
+        }
+    }
+}
+
+/// Create a retry stream that will resubscribe on disconnect.
+///
+/// It's important to note that this function is intended to work only for stateless subscriptions.
+/// If the subscription takes input or modifies state, this function should not be used.
+pub(crate) async fn retry_stream<F, R>(mut sub_stream: F) -> Result<StreamOfResults<R>, Error>
+where
+    F: FnMut() -> ResubscribeFuture<R> + Send + 'static + Clone,
+    R: Send + 'static,
+{
+    loop {
```

Comment:
It's a shame that we are sortof duplicating the try logic in an outer and then inner loop here, I think just to return the initial error in the `Result` before we are inside a stream and errors come in the streams.

I wonder if it makes sense to change the signature of this fn to return `-> StreamOfResults<R>` (ie no outer `Result`) and then the body can be simpler eg:

```rust
// future that will resubscribe to the stream, internally retrying 
// if any DisconnectedWillReconnect (I wonder whether `retry` could be 
// used to make this bit even shorter too..)
let resubscribe = Box::new(move || {...});

// The extra Box is to encapsulate the retry subscription type
return Ok(StreamOf::new(Box::pin(RetrySubscription {
    state: PendingOrStream::Stream(v),
    resubscribe,
})));
```

Reply:
Maybe best to leave as is for this PR though because changing it above might hjave knock-on effects on the signatures of backend fns or whatever.

Reply:
I think you will be happier now after I removed the !DRY stuff